### PR TITLE
ENH: Add SIMD versions of bool logical_&&,||,! and absolute

### PR DESCRIPTION
**Pull request by Developer-Ecosystem-Engineering**
NumPy has SIMD versions of BOOL `logical_and`, `logical_or`, `logical_not`, and `absolute` for SSE2.  The changes here replace that implementation with one that uses universal intrinsics.  This allows other architectures to have SIMD versions of the functions too.

BOOL `logical_and` and `logical_or` are particularly important for NumPy as that's how  `np.any()` / `np.all()` are implemented.

Apple M1: **up to 16.5x faster**
```
       before           after         ratio
     [7c143834]       [c49d9dc2]
     <main>           <logical/dev>
-        3.47±0μs      1.82±0.01μs     0.52  bench_reduce.AnyAll.time_all_slow
-     4.05±0.01μs      1.83±0.06μs     0.45  bench_reduce.AnyAll.time_any_slow
-        6.55±0μs          532±2ns     0.08  bench_ufunc.Custom.time_not_bool
-        10.2±0μs          680±7ns     0.07  bench_ufunc.Custom.time_or_bool
-     11.0±0.07μs          665±3ns     0.06  bench_ufunc.Custom.time_and_bool
                                                                                                                                                                    
SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

Apple M1 (Rosetta): **up to 1.6x faster**
```
       before           after         ratio
     [7c143834]       [1ad2f2f2]
     <main>           <logical/dev>
-     1.14±0.01μs         1.03±0μs     0.90  bench_ufunc.Custom.time_not_bool
-     4.38±0.03μs      3.06±0.02μs     0.70  bench_reduce.AnyAll.time_any_slow
-     5.20±0.01μs      3.17±0.01μs     0.61  bench_reduce.AnyAll.time_all_slow

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

iMac Pro (AVX512): **up to 1.2x faster**
```
       before           after         ratio
     [da6297b9]       [c49d9dc2]
     <main>           <logical/dev>
-     1.42±0.02μs      1.24±0.03μs     0.87  bench_ufunc.Custom.time_and_bool
-     4.16±0.03μs       3.60±0.1μs     0.86  bench_reduce.AnyAll.time_any_slow
-     1.21±0.03μs      1.03±0.02μs     0.86  bench_ufunc.Custom.time_not_bool
-      4.30±0.1μs      3.56±0.07μs     0.83  bench_reduce.AnyAll.time_all_slow

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

### CODE REVIEW

File: numpy/core/setup.py.orig
Comment (mattip) [inline numpy/core/setup.py.orig:1]: This backup of `setup.py` should be removed.

**Review by mattip — COMMENTED** [review]
Nice. I will defer to @seiko2plus for the deeper SIMD review. A couple of general questions for all the latest SIMD PRs:
- could you check the binary file size change of `_multiarray_umath*.so` after vs. before?
- Did the change have an effect on any other benchmarks besides the obvious ones that you reported
- If possible, could you run the benchmarks on a non-avx512 x86_64 system

File: numpy/core/src/umath/loops_logical.dispatch.c.src
Comment (seiko2plus) [inline numpy/core/src/umath/loops_logical.dispatch.c.src:51]:
```suggestion
```
covered by #22306, in case if still required.

File: numpy/core/src/umath/loops_logical.dispatch.c.src
Comment (seiko2plus) [inline numpy/core/src/umath/loops_logical.dispatch.c.src:24]:
```suggestion
```
no need for two paths, new intrinsics any/all covered by #22306 should provide fair optimization across all SIMD extensions

**Review by seiko2plus — CHANGES_REQUESTED** [review]
Sorry for the delayed response. For fair optimization and to avoid two separate implementations for reduction, new universal intrinsic any/all going to be implemented through pr #22306.

**Comment by Developer-Ecosystem-Engineering** [comment]
I merged in the suggestions made to adopt changes in #22306 (didn't exist at the time of this PR authorship)

File: numpy/core/setup.py.orig
Comment (seiko2plus) [reply numpy/core/setup.py.orig:1]: still

File: numpy/core/src/umath/loops_logical.dispatch.c.src
Comment (seiko2plus) [inline numpy/core/src/umath/loops_logical.dispatch.c.src:127]:
```suggestion
        if ([! , ] npyv_[all, any]_u8(v0)) {
```
should provide better performance across all SIMD extensions

File: numpy/core/src/umath/loops_logical.dispatch.c.src
Comment (seiko2plus) [inline numpy/core/src/umath/loops_logical.dispatch.c.src:200]:
```suggestion
```

**Review by seiko2plus — CHANGES_REQUESTED** [review]
Excellent work, just one nit more.

**Review by seiko2plus — CHANGES_REQUESTED** [review]
My apologies, It seems I wasn't entirely focused. One thing more.

**Review by seiko2plus — APPROVED** [review]
LGTM, Thank you.
I have made some additional changes to umath generator to enable runtime dispatching for the following alias:
```
BOOL_invert, BOOL_add, BOOL_bitwise_and
BOOL_bitwise_or, BOOL_logical_xor
BOOL_bitwise_xor, BOOL_multiply
BOOL_maximum, BOOL_minimum, BOOL_fmax,
BOOL_fmin
```

**Comment by seiko2plus** [comment]
The last push after the approval to satisfy python linter and to fix dispatch karg on umath generator for `logical_xor`, both errors were caused by my latest patch.

**Comment by Developer-Ecosystem-Engineering** [comment]
@mattip Anything else missing/required?

**Comment by mattip** [comment]
Thanks @Developer-Ecosystem-Engineering and @seiko2plus

**Comment by tylerjereddy** [comment]
Note that this showed up as a bit naughty in `git bisect` from gh-22845.

**Pull request merged**
